Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp

```diff
@@ -990,13 +988,138 @@ void WebAutomationSession::fragmentNavigatedForFrame(const WebFrameProxy& frame,
+        if (RefPtr openerFrame = WebFrameProxy::webFrame(*openerFrameID))
+            return openerFrame->page();
+    }
+    return nullptr;
+}
+
+void WebAutomationSession::contextCreatedForFrame(const WebFrameProxy& frame)
+{
+    auto contextHandle = handleForWebFrameProxy(frame);
+    auto url = frame.url().string();
+    auto parentHandle = frame.parentFrame() ? handleForWebFrameProxy(*frame.parentFrame()) : "null"_s;
+
+    String clientWindow = "unknown-window"_s;
+    String userContext = "default"_s;
+
+    if (RefPtr page = frame.page()) {
+        auto [windowId, contextId] = getClientWindowAndUserContext(*page);
+        clientWindow = windowId;
+        userContext = contextId;
+    }
+
+    m_bidiProcessor->browsingContextDomainNotifier().contextCreated(contextHandle, url, "null"_s, parentHandle, JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>::create(), clientWindow, userContext);
+}
+
+void WebAutomationSession::recursivelyEmitContextCreated(const FrameTreeNodeData& tree, std::optional<String>&& parentContext)
+{
+    RefPtr frame = WebFrameProxy::webFrame(tree.info.frameID);
+    if (!frame)
+        return;
+
+    RefPtr page = frame->page();
+    if (!page)
+        return;
+
+    String contextHandle;
+    String originalOpener = "null"_s;
+
+    if (tree.info.isMainFrame) {
+        contextHandle = handleForWebPageProxy(*page);
+        if (RefPtr openerPage = getOpenerPage(*page))
+            originalOpener = handleForWebPageProxy(*openerPage);
+    } else
+        contextHandle = handleForWebFrameID(tree.info.frameID);
+
+    String url = tree.info.request.url().string();
+    auto [clientWindow, userContext] = getClientWindowAndUserContext(*page);
+    auto children = JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>::create();
+
+    m_bidiProcessor->browsingContextDomainNotifier().contextCreated(contextHandle, url, originalOpener, parentContext.value_or("null"_s), WTFMove(children), clientWindow, userContext);
+
+    for (const auto& child : tree.children)
+        recursivelyEmitContextCreated(child, contextHandle);
+}
+
+void WebAutomationSession::contextDestroyedForPage(const WebPageProxy& page)
+{
+    auto contextHandle = handleForWebPageProxy(page);
+    auto url = page.currentURL();
+
+    String parentContext = "null"_s;
+    if (RefPtr mainFrame = page.mainFrame(); mainFrame && mainFrame->parentFrame())
+        parentContext = handleForWebFrameProxy(*mainFrame->parentFrame());
+
+    String originalOpener = "null"_s;
+    if (RefPtr openerPage = getOpenerPage(page))
+        originalOpener = handleForWebPageProxy(*openerPage);
+
+    auto [clientWindow, userContext] = getClientWindowAndUserContext(page);
+
+    m_bidiProcessor->browsingContextDomainNotifier().contextDestroyed(contextHandle, url, originalOpener, parentContext, JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>::create(), clientWindow, userContext);
+
+    m_handleWebPageMap.remove(contextHandle);
```

Comment:
I think it's OK to do this map maintenance here, but then we need to ensure that these two new instrumentation methods are called unconditionally. As-is, in this patch we remove some bookkeeping but don't add it back if `!ENABLE(WEBDRIVER_BIDI)`